src/srb2kart/serverinfo/serverinfo.go

```diff
@@ -0,0 +1,297 @@
+	Header         KartPacketHeader
+	PacketVersion  uint8
+	Application    [16]byte
+	Version        uint8
+	SubVersion     uint8
+	NumberOfPlayer uint8
+	MaxPlayers     uint8
+	GameType       uint8
+	ModifiedGame   bool
+	CheatsEnabled  bool
+	KartVars       uint8
+	FileNeededNum  uint8
+	Time           uint32
+	LevelTime      uint32
+	ServerName     string
+	MapName        string
+	MapTitle       string
+	MapMD5         [16]uint8
+	ActNum         uint8
+	IsZone         uint8
+	HttpSource     string
+	FileNeeded     []FileNeededEntry
+}
+
+func (p *KartServerInfoPacket) GetPacketType() packetType {
+	return p.Header.PacketType
+}
+
+type FileNeededEntry struct {
+	WillSend bool
+	TotalSize uint32
+	FileName string
+	MD5 [16]uint8
+}
+
+type kartPlayerInfoEntryRaw struct {
+	Node         uint8
+	Name         [21 + 1]byte
+	Address      [4]uint8
+	Team         uint8
+	Skin         uint8
+	Data         uint8
+	Score        uint32
+	TimeInServer uint16
+}
+
+type kartPlayerInfoPacketRaw struct {
+	Header     KartPacketHeader
+	PlayerInfo [32]kartPlayerInfoEntryRaw
+}
+
+type KartPlayerInfoEntry struct {
+	Node         uint8
+	Name         string
+	Address      [4]uint8
+	Team         uint8
+	Skin         uint8
+	Data         uint8
+	Score        uint32
+	TimeInServer uint16
+}
+
+type KartPlayerInfoPacket struct {
+	Header     KartPacketHeader
+	PlayerInfo []KartPlayerInfoEntry
+}
+
+func (p *KartPlayerInfoPacket) GetPacketType() packetType {
+	return p.Header.PacketType
+}
+
+func nullTerminated(data []byte) string {
+	newBytes := make([]byte, 0)
+
+	// Filter out all bytes above 127
+	for i := 0; i < len(data); i++ {
+		if data[i] <= 127 {
+			newBytes = append(newBytes, data[i])
+		}
+	}
+	before, _, _ := bytes.Cut(newBytes, []byte{0})
+	return string(before)
+}
+
+func readPacket(data []byte) (KartPacket, error) {
+	header := KartPacketHeader{}
+	if err := binary.Read(bytes.NewReader(data), binary.LittleEndian, &header); err != nil {
+		return nil, err
+	}
+
+	switch header.PacketType {
+	case packetTypeServerInfo:
+		return unpackServerInfoPacket(data)
+	case packetTypePlayerInfo:
+		return unpackPlayerInfoPacket(data)
+	default:
+		return nil, fmt.Errorf("unknown packet type: %d", header.PacketType)
+	}
+}
+
+func parseFileNeeded(data [915]byte, fileNeededCount int) []FileNeededEntry {
+	var entries []FileNeededEntry
+	buf := bytes.NewBuffer(data[:])
+	for i := 0; i < fileNeededCount; i++ {
+		entry := FileNeededEntry{}
+		binary.Read(buf, binary.LittleEndian, &entry.WillSend)
+		b, _ := buf.ReadByte()
+		entry.WillSend = (b >> 4) == 1
+		binary.Read(buf, binary.LittleEndian, &entry.TotalSize)
+		entry.FileName, _ = buf.ReadString(0)
+		// Trim off the null terminator
+		entry.FileName = entry.FileName[:len(entry.FileName)-1]
+		binary.Read(buf, binary.LittleEndian, &entry.MD5)
+		entries = append(entries, entry)
+	}
+	return entries
+}
+
+func unpackServerInfoPacket(data []byte) (*KartServerInfoPacket, error) {

```

Comment:
Imo if the only reason to return a pointer is to be able to return nil on error, while the function also constructs the type, it's better to return the value and on error return an uninitialised value.